File: Example/Tests/MediaControlTests.m
//
//  MediaControlTests.m
//  Clappr
//
//  Created by Gustavo Barbosa on 12/18/14.
//  Copyright (c) 2014 globo.com. All rights reserved.
//

#import <Clappr/Clappr.h>
#import <objc/runtime.h>

SPEC_BEGIN(MediaControl)

describe(@"Media Control", ^{

    __block CLPMediaControl *mediaControl;
    __block CLPContainer *container;
    __block CLPPlayback *playback;

    NSURL *sourceURL = [NSURL URLWithString:@"http://globo.com/video.mp4"];

    beforeEach(^{
        playback = [[CLPPlayback alloc] initWithURL:sourceURL];
        container = [[CLPContainer alloc] initWithPlayback:playback];
        mediaControl = [[CLPMediaControl alloc] initWithContainer:container];
    });

    describe(@"Play", ^{

        it(@"should contain a play/pause button embed in its view", ^{
            UIButton *playPauseButton = mediaControl.playPauseButton;
            [[playPauseButton.superview should] equal:mediaControl.view];
        });

        it(@"should be triggered after touch the button when it is not playing", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(NO)];

            [[mediaControl should] receive:@selector(play)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should call container play after its play method has been called", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(NO)];

            [[container should] receive:@selector(play)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should trigger playing event", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(NO)];

            __block BOOL callbackWasCalled = NO;
            [mediaControl once:CLPMediaControlEventPlaying callback:^(NSDictionary *userInfo) {
                callbackWasCalled = YES;
            }];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];

            [[theValue(callbackWasCalled) should] beTrue];
        });
    });

    describe(@"Pause", ^{

        it(@"should be triggered after touch the button when it is playing", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(YES)];

            [[mediaControl should] receive:@selector(pause)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });

        it(@"should call container pause after its pause method has been called", ^{

            [container stub:@selector(isPlaying) andReturn:theValue(YES)];

            [[container should] receive:@selector(pause)];

            [mediaControl.playPauseButton sendActionsForControlEvents:UIControlEventTouchUpInside];
        });
    });

    describe(@"Event listening", ^{

        it(@"should toggle play after listen to container's play event", ^{
            // change the view
        });

        it(@"should trigger media control's playing event after listen to container's play event", ^{

            __block BOOL callbackWasCalled = NO;
            [mediaControl once:CLPMediaControlEventPlaying callback:^(NSDictionary *userInfo) {
                callbackWasCalled = YES;
            }];

            [container trigger:CLPContainerEventPlay];

            [[theValue(callbackWasCalled) should] beTrue];
        });

        it(@"should trigger media control's not playing event after listen to container's pause event", ^{

            __block BOOL callbackWasCalled = NO;
            [mediaControl once:CLPMediaControlEventNotPlaying callback:^(NSDictionary *userInfo) {
                callbackWasCalled = YES;
            }];

            [container trigger:CLPContainerEventPause];

            [[theValue(callbackWasCalled) should] beTrue];
        });

        it(@"should update seek bar after listen to container's time update event", ^{

        });

        it(@"should update progress bar after listen to container's progress event", ^{

        });

        it(@"should handle container's settings update event properly", ^{

        });

        it(@"should handle container's DVR state event properly ", ^{

        });

        it(@"should handle container's HD update properly", ^{

        });

        it(@"should disable after listen to container's media control disable event", ^{

        });

        it(@"should enable after listen to container's media control enable event", ^{

        });

        it(@"should handle container's ended event properly", ^{

        });
    });
});

SPEC_END

File: Example/Tests/MediaControlTests.swift
import Quick
import Nimble
import Clappr

class MediaControlTests: QuickSpec {
    
    override func spec() {
        describe("MediaControl") {
            let sourceUrl = NSURL(string: "http://globo.com/video.mp4")!
            var container: StubedContainer!
            
            beforeEach() {
                container = StubedContainer(playback: Playback(url: sourceUrl))
            }
            
            context("Initialization") {
                
                it("Should have a init method that receives a container") {
                    let mediaControl = MediaControl.initWithContainer(container)
                    
                    expect(mediaControl).toNot(beNil())
                    expect(mediaControl.container) == container
                }
            }
            
            context("Behavior") {
                var mediaControl: MediaControl!
                
                beforeEach() {
                    mediaControl = MediaControl.initWithContainer(container)
                }
                
                context("Visibility") {
                    it("Should start with controls visible") {
                        expect(mediaControl.playPauseButton.hidden).to(beFalse())
                        expect(mediaControl.controlsOverlayView.hidden).to(beFalse())
                        expect(mediaControl.controlsWrapperView.hidden).to(beFalse())
                    }
                    
                    it("Should hide it's control after hide is called") {
                        mediaControl.hide()
                        
                        expect(mediaControl.playPauseButton.hidden).to(beTrue())
                        expect(mediaControl.controlsOverlayView.hidden).to(beTrue())
                        expect(mediaControl.controlsWrapperView.hidden).to(beTrue())
                    }
                    
                    it("Should show it's control after show is called") {
                        mediaControl.hide()
                        mediaControl.show()
                        
                        expect(mediaControl.playPauseButton.hidden).to(beFalse())
                        expect(mediaControl.controlsOverlayView.hidden).to(beFalse())
                        expect(mediaControl.controlsWrapperView.hidden).to(beFalse())
                    }
                }
                
                context("Play") {
                    it("Should call container play") {
                        mediaControl.playPauseButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(container.playWasCalled).to(beTrue())
                        expect(container.pauseWasCalled).to(beFalse())
                    }
                    
                    it("Should change button state to selected") {
                        expect(mediaControl.playPauseButton.selected).to(beFalse())
                        mediaControl.playPauseButton.sendActionsForControlEvents(UIControlEvents.TouchUpInside)
                        expect(mediaControl.playPauseButton.selected).to(beTrue())
                    }
                }
            }
        }
    }
    
    class StubedContainer: Container {
        var playWasCalled = false
        var pauseWasCalled = false
        
        override func play() {
            playWasCalled = true
        }
        
        override func pause() {
            pauseWasCalled = true
        }
    }
}